### PR TITLE
Better default setup on `MoneyField` for models

### PR DESCRIPTION
**Pull request by flaeppe**
Hi, I was digging in to making some progress on #638 (refs: #621) and realised that there was some adjustments that could be made for defaults on `djmoney.models.fields.MoneyField`. I've gathered several (minor) fixes in this PR(see list below), all of them originating from the same method.

- Removes possibility of allowing _any_ currency code on string based default value
- Having a default as type `bytes` is now properly supported
- Asserts on incompatible defaults with default currency being `None` when amount has a valid type (i.e. a default setup generating `Money(10, None)`)

_I think_ the changes here pushes work for #638 at least a little bit forward, as one of the major changes there would (probably?) be setting `DEFAULT_CURRENCY=None` per default. While the changes here help out on management of `MoneyField(..., default_currency=None)`. And I was thinking it could be nice taking the opportunity helping reduce the size of the work on #638.

Anyways, let me know what you think.

### CODE REVIEW

File: djmoney/models/fields.py [hunk 1]
```diff
@@ -196,25 +196,34 @@ def __init__(
         Field.creation_counter += 1
 
     def setup_default(self, default, default_currency, nullable):
-        if isinstance(default, (str, bytes)):
+        if default in (None, NOT_PROVIDED) or isinstance(default, Money):
```
Comment (Stranger6667) [inline djmoney/models/fields.py:199]: Should it be `MONEY_CLASSES` in `isinstance` as in other places above?

Comment (flaeppe) [reply djmoney/models/fields.py:199]: I don't think so, as a bit further down we'll convert `OldMoney -> Money`

https://github.com/flaeppe/django-money/blob/e7d558c9cdf92227e997ef65fb6f1d3cdf87c14b/djmoney/models/fields.py#L216-L217

**Comment by flaeppe** [comment]
I'm not sure why CI runs with `pypy3` now has an sqlite version that's incompatible with `django`.

**Comment by flaeppe** [comment]
> I'm not sure why CI runs with `pypy3` now has an sqlite version that's incompatible with `django`.

I think the closest I got to is somewhere here: https://stackoverflow.com/questions/65476852/make-pypy-in-virtualenv-use-shared-library-from-os-rather-than-its-own-copy-lib

No idea if deleting PyPy's copies of `libsqlite3.so` would actually help though..

**Comment by flaeppe** [comment]
@Stranger6667 any insight on why running with `pypy` would fail on the sqlite version? Do you know if we could do something to resolve it?

Additionally, there's an odd error (new to me at least) on the `pre-commit` job. Can be seen here: https://github.com/django-money/django-money/runs/5633124684?check_suite_focus=true . Not sure if rerunning the workflow would avoid it

**Comment by flaeppe** [comment]
> @Stranger6667 any insight on why running with `pypy` would fail on the sqlite version? Do you know if we could do something to resolve it?
> 
> Additionally, there's an odd error (new to me at least) on the `pre-commit` job. Can be seen here: https://github.com/django-money/django-money/runs/5633124684?check_suite_focus=true . Not sure if rerunning the workflow would avoid it

Seems that it was fixed with `actions/setup-python` action bumping pypy version for Ubuntu20 (https://github.com/actions/virtual-environments/commit/1ede8085f53bf7b38163389282bed3b8173b4d62)

**Review by benjaoming — APPROVED** [review]
Solid work :100: 

If @Stranger6667 isn't around for another review, I'll get it in within the next days - meanwhile, release notes for this are more than welcome :)

File: tests/test_models.py [hunk 2]
```diff
@@ -302,6 +304,80 @@ def test_null_default(self):
         instance = NullMoneyFieldModel.objects.create()
         assert instance.field is None
 
+    def test_implicit_currency_field_not_nullable_when_money_field_not_nullable(self):
+        instance = ModelWithVanillaMoneyField(money=10, money_currency=None)
+        with pytest.raises(CurrencyDoesNotExist, match=r"No currency with code NONE"):
```
Comment (Stranger6667) [inline tests/test_models.py:309]: I think it would be better to avoid implicit conversion to string and provide a more specific error message for `None`.

**Review by Stranger6667 — COMMENTED** [review]
Thanks for pinging me @benjaoming :)

I have a small comment on one of the error messages

File: tests/test_models.py (hunk 2 above)
Comment (flaeppe) [reply tests/test_models.py:309]: I think I see what you're getting at, if my 5 cents are worth anything, I'm of the opinion that's within the domain of `py-moneyed`(?). These are the lines where `"NONE”` would originate from:

https://github.com/py-moneyed/py-moneyed/blob/c495642244f629e4ed330b25438edbd517175146/src/moneyed/classes.py#L193-L197

`None` would get cast to `"NONE"` before it's validated as a currency code.

In addition to that, if we look at the `__init__` signature of `class Money` in `py-moneyed` (that runs the linked code above) it doesn't define any support for passing `currency=None`;

https://github.com/py-moneyed/py-moneyed/blob/c495642244f629e4ed330b25438edbd517175146/src/moneyed/classes.py#L173-L179

https://github.com/py-moneyed/py-moneyed/blob/c495642244f629e4ed330b25438edbd517175146/src/moneyed/classes.py#L186-L189

As such, I'm leaning towards the interpretation that `None` _is already an_ unsupported value for a currency code, therefore resulting in undefined behaviour (although kindly enough with the version of `py-moneyed` that's used we get an error with `"NONE"` attempted as currency code).

---

I'm aware of that I linked from `HEAD` at `py-moneyed`, but even if we're using an older version of `py-moneyed` which defines this differently. I kind of don't think it's worth to duplicate the code to handle `None`:

https://github.com/py-moneyed/py-moneyed/blob/c495642244f629e4ed330b25438edbd517175146/src/moneyed/classes.py#L186-L189

But _if_ we should do something different for `None`, I think we should "backport" those lines, throwing a `TypeError`.

Comment (flaeppe) [reply tests/test_models.py:309]: I also can't remember why I added this test in the first place. Might have been just for coverage or so..

Comment (benjaoming) [reply tests/test_models.py:309]: Isn't the AssertionError raised in L221 what would be the expected outcome of this test case?

File: djmoney/models/fields.py [hunk 3]
```diff
@@ -196,25 +196,34 @@ def __init__(
...
-        return default
+            amount, currency = default.amount, default.currency
+        else:
+            raise ValueError(f"default value must be an instance of Money, is: {default}")
+
+        assert currency is not None, (
+            "Default currency can not be `None` when default is of a value that doesn't include a currency. Either"
+            " provide a default containing a currency value or configure a currency default via setting"
+            " `default_currency=` or `DEFAULT_CURRENCY`(global)"
+        )
```
Comment (benjaoming) [inline djmoney/models/fields.py:225]: Is this part reachable in tests? And should it perhaps also figure as a ValueError?

Comment (flaeppe) [reply djmoney/models/fields.py:225]: Yeah, it's reachable. It's covered by these tests:

https://github.com/django-money/django-money/blob/6dcdbeb685731ebbcf2d38638b3f6f87e21e6f1e/tests/test_models.py#L316-L320

My idea behind raising an `AssertionError` was to signal that it's an exception that's not intended to be caught by the caller.

File: tests/test_models.py (hunk 2 above)
Comment (flaeppe) [reply tests/test_models.py:309]: Not really, since that line is for initiating a `MoneyField` on a model(default value for currency), while this is setting a value for an already initiated `MoneyField`.

Comment (flaeppe) [reply tests/test_models.py:309]: > I think it would be better to avoid implicit conversion to string and provide a more specific error message for `None`.

I don't have too strong opinions handling `None` in django-money and raising a `TypeError`, if that's what you guys think is best.

Let me know and I'll bump the PR.

File: djmoney/models/fields.py (hunk 3 above)
Comment (flaeppe) [reply djmoney/models/fields.py:225]: > My idea behind raising an AssertionError was to signal that it's an exception that's not intended to be caught by the caller.

While thinking twice about that, it'd be quite weird for someone to actually catch that exception anyways:

```python
class MyModel(models.Model):
    try:
        money = MoneyField(...)
    except ValueError:
        # ??
```

If you'd prefer a `ValueError` there, I'm fine with changing it.

Comment (benjaoming) [reply djmoney/models/fields.py:225]: I understand your great explanation and rationale -- and no, this is not at all important to entertain going deeper into!

Comment (benjaoming) [reply djmoney/models/fields.py:225]: Thanks for explaining!

File: tests/test_models.py (hunk 2 above)
Comment (benjaoming) [reply tests/test_models.py:309]: I'm happy that wrong input (perhaps an edge case) is blocked in a nice way that will guide the developer to solve the issue, so I think we can move on :+1:

Comment (flaeppe) [reply tests/test_models.py]: I changed it to raise a `TypeError` just now.

Comment (benjaoming) [reply tests/test_models.py]: @flaeppe - thanks for updating the solution!

@Stranger6667 - to me, this looks good and fine, does the elaboration and the update work for you?

Comment (Stranger6667) [reply tests/test_models.py]: Sounds good to me!

Comment (Stranger6667) [reply tests/test_models.py]: Thanks for all the details :)

**Comment by benjaoming** [comment]
Thanks for the patience @flaeppe :pray: :hourglass_flowing_sand:  :rocket:
